File: java/longestRepeatingCharacter.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character, Integer> characterCount = new HashMap();
        char[] charArray = s.toCharArray();
        int maxValid = 0;
        int maxOccurrence = 0;
        int left = 0;


        for (int right = 0; right < charArray.length; right++) {
            Character currentChar = charArray[right];
            int newCount = 1;

            if (characterCount.containsKey(currentChar)) {
                newCount = characterCount.get(currentChar) + newCount;
                characterCount.put(currentChar, newCount);
            } else {
                characterCount.put(currentChar, newCount);
            }

            maxOccurrence = Math.max(maxOccurrence, newCount);

            // if invalid, move left pointer right
            while (((right - left) + 1) - maxOccurrence > k) {
                Character leftChar = charArray[left];
                int prevLeft = characterCount.get(leftChar);
                characterCount.put(leftChar, prevLeft - 1);

                left++;
            };

            maxValid = Math.max(maxValid, (right - left) + 1);
        }

        return maxValid;
    };
}

File: js/longestRepeatingCharacter.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
const characterReplacement = function(s, k) {
    let left = 0;
    let maxOccurence = 0;
    const characterCount = {}
    let maxValid = 0;

    for (let right = 0; right < s.length; right++) {
        const newCharCount = (characterCount[s[right]] ?? 0) + 1
        characterCount[s[right]] = newCharCount

        maxOccurence = Math.max(maxOccurence, newCharCount);

        // if window is invalid, shrink from the left
        // if window length - maxOccurence <= k
        while (((right + 1) - left) - maxOccurence > k) {
            characterCount[s[left]]--
            left++;
        }

        maxValid = Math.max(maxValid, (right + 1) - left)
    }

    return maxValid;
};